translate: basic/*-sorts (#124)
OI-wiki/i18n: docs/basic/stl-sort.md

However the C++ standard does not strictly specify the implementation of this function and it depends on the compiler used. In both [libstdc++](https://github.com/mirrors/gcc/blob/master/libstdc++-v3/include/bits/stl_algo.h) and [libc++](http://llvm.org/svn/llvm-project/libcxx/trunk/include/algorithm), [introsort](https://en.wikipedia.org/wiki/Introsort) is used.

## std::nth_element

Reference: [`std::nth_element`](https://en.cppreference.com/w/cpp/algorithm/nth_element)

Usage:

```cpp
std::nth_element(first, nth, last);
std::nth_element(first, nth, last, cmp);
```

It re-sorts elements between $[first, last)$, making the element pointed by $nth$ be the element at the same position in ordered version of $[first, last)$. Any elements before the new $n$-th element are equal to or smaller than the new $n$-th element.

The implementation is the unfinished introsort.

For the two usages, the C++ standard requires their average-case time complexity to be $O(n)$, where `n = std::distance(first, last)`.

> Often used to build [K-D Tree](https://en.wikipedia.org/wiki/K-d_tree). <!---暂时保留-->

Reviewer: ```suggestion
Often used to build [K-D Tree](https://en.wikipedia.org/wiki/K-d_tree).
```

std::stable_sort(begin, end, cmp);
```

Stable sort means that the relative positions of equal elements after the sorting are the same as the original sequence.

The time complexity is $O(n\log (n)^2)$, or $O(n\log n)$ if extra memory is available.

## std::partial_sort

Reference: [`std::partial_sort`](https://en.cppreference.com/w/cpp/algorithm/partial_sort)

Usage:

```cpp
std::partial_sort(first, mid, last);
std::partial_sort(first, mid, last, cmp);
```

Place the top $k$ smallest elements in the first $k$ positions in the sequence (the order of equal elements is not guaranteed), and the order of the remaining elements is not specified. If `cmp` is not specified the array will be sorted from smallest to largest by default (or, `std::less<>`). 

> Time complexity: Approximately performing $(last-first)\log(mid-first)$ times `cmp()`.

Reviewer: ```suggestion
Time complexity: Approximately performing `cmp()` for $(last-first)\log(mid-first)$ times.
```

感觉 times 有些时候会被误解为乘号

Author: 嗯嗯~~那就尽量避免出现

Author: will fix manually

  return (u1.a == u2.a) ? (u1.b > u2.b) : (u1.a > u2.a);
}
// ......
std::sort(da + 1, da + 1 + 10);  // use the <operator defined in the structure to sort in ascending order
std::sort(da + 1, da + 1 + 10, cmp);  // use cmp function to compare, sort in descending order
```

### Strictly weak order

> Reference: [Strict weak orderings](https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings)

Reviewer: 参见是 See Also？

Author: ...a

Author: #FixManually

> 
> 

Reviewer: ```suggestion

```

- Use `<=` to define the less-than operator in sorting.
- When calling the sorting operator, reading the external value may change the array. (commonly seen in the shortest path algorithm)
- The result of comparing the maximum and minimum values of multiple numbers is used as a sorting operator. (For example, the classic error in queens game or processing production scheduling)

## External Links

> - [Talking about the application of neighbor exchange sorting and problems that need attention](https://ouuan.github.io/浅谈邻项交换排序的应用以及需要注意的问题/) (Original link in Chinese).

Reviewer: ```suggestion
- [Discussing the application of neighbor exchange sorting and problems that need attention](https://ouuan.github.io/浅谈邻项交换排序的应用以及需要注意的问题/) (Original link in Chinese).
```

Author: 啊咱的本意是 a talk about

Reviewer: 🤔 A talk 的话做自己写的博客/文章标题不太适合？
